File: common-js-lib/templates/ServiceClientTemplate.ts
/**
 * Standardized Service Client Template for TypeScript
 * 
 * This module provides a template for creating service clients that follow the
 * platform's standardized patterns for service communication, error handling,
 * resilience, and metrics collection.
 * 
 * Usage:
 * 1. Copy this template to your service's clients directory
 * 2. Rename the class to match your service (e.g., MarketDataClient)
 * 3. Implement service-specific methods using the base HTTP methods
 * 4. Create a factory function in your service's clientFactory.ts
 * 
 * Example:
 * ```typescript
 * import { BaseServiceClient, ClientConfig } from 'common-js-lib';
 * 
 * export class ExampleServiceClient extends BaseServiceClient {
 *   constructor(config: ClientConfig) {
 *     super(config);
 *   }
 *   
 *   async getResource(resourceId: string): Promise<any> {
 *     return this.get(`resources/${resourceId}`);
 *   }
 * }
 * ```
 */

import { 
  BaseServiceClient, 
  ClientConfig, 
  ClientError,
  ClientConnectionError,
  ClientTimeoutError,
  ClientValidationError,
  ClientAuthenticationError
} from '../index';

/**
 * Standardized service client template
 * 
 * This class extends BaseServiceClient with additional standardized methods
 * and patterns for service communication.
 * 
 * Features:
 * 1. Consistent error handling and logging
 * 2. Type safety with generics
 * 3. Standardized method signatures
 * 4. Built-in request/response logging
 * 5. Correlation ID propagation
 */
export class StandardServiceClient<T = any> extends BaseServiceClient {
  /**
   * Initialize the standardized service client
   * 
   * @param config Client configuration
   */
  constructor(config: ClientConfig) {
    super(config);
  }
  
  /**
   * Template method for getting a resource by ID
   * 
   * @param resourceId Resource identifier
   * @returns Resource data
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async getResource(resourceId: string): Promise<T> {
    this.logger.debug(`Getting resource ${resourceId}`);
    try {
      return await this.get(`resources/${resourceId}`);
    } catch (error) {
      this.logger.error(`Failed to get resource ${resourceId}: ${error}`);
      throw error;
    }
  }
  
  /**
   * Template method for creating a resource
   * 
   * @param data Resource data
   * @returns Created resource data
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async createResource(data: any): Promise<T> {
    this.logger.debug(`Creating resource`);
    try {
      return await this.post('resources', data);
    } catch (error) {
      this.logger.error(`Failed to create resource: ${error}`);
      throw error;
    }
  }
  
  /**
   * Template method for updating a resource
   * 
   * @param resourceId Resource identifier
   * @param data Updated resource data
   * @returns Updated resource data
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async updateResource(resourceId: string, data: any): Promise<T> {
    this.logger.debug(`Updating resource ${resourceId}`);
    try {
      return await this.put(`resources/${resourceId}`, data);
    } catch (error) {
      this.logger.error(`Failed to update resource ${resourceId}: ${error}`);
      throw error;
    }
  }
  
  /**
   * Template method for deleting a resource
   * 
   * @param resourceId Resource identifier
   * @returns Deletion confirmation
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async deleteResource(resourceId: string): Promise<any> {
    this.logger.debug(`Deleting resource ${resourceId}`);
    try {
      return await this.delete(`resources/${resourceId}`);
    } catch (error) {
      this.logger.error(`Failed to delete resource ${resourceId}: ${error}`);
      throw error;
    }
  }
  
  /**
   * Template method for listing resources
   * 
   * @param filters Optional filters to apply
   * @param page Page number
   * @param pageSize Number of items per page
   * @returns List of resources
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async listResources(filters?: Record<string, any>, page: number = 1, pageSize: number = 100): Promise<{ data: T[], meta: any }> {
    this.logger.debug(`Listing resources (page=${page}, pageSize=${pageSize})`);
    const params: Record<string, any> = { page, pageSize };
    
    if (filters) {
      Object.assign(params, filters);
    }
    
    try {
      return await this.get('resources', { params });
    } catch (error) {
      this.logger.error(`Failed to list resources: ${error}`);
      throw error;
    }
  }
  
  /**
   * Template method for executing a custom operation
   * 
   * @param operation Operation name
   * @param resourceId Optional resource identifier
   * @param data Optional operation data
   * @returns Operation result
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   * @throws ClientValidationError if the request is invalid
   * @throws ClientAuthenticationError if authentication fails
   */
  async executeOperation(operation: string, resourceId?: string, data?: any): Promise<any> {
    const endpoint = resourceId ? `resources/${resourceId}/${operation}` : `operations/${operation}`;
    
    this.logger.debug(`Executing operation ${operation}`);
    try {
      return await this.post(endpoint, data);
    } catch (error) {
      this.logger.error(`Failed to execute operation ${operation}: ${error}`);
      throw error;
    }
  }
  
  /**
   * Check the health of the service
   * 
   * @returns Health status
   * @throws ClientError if the request fails
   * @throws ClientConnectionError if connection to the service fails
   * @throws ClientTimeoutError if the request times out
   */
  async getHealth(): Promise<any> {
    this.logger.debug(`Checking service health`);
    try {
      return await this.get('health');
    } catch (error) {
      this.logger.error(`Failed to check service health: ${error}`);
      throw error;
    }
  }
  
  /**
   * Create a new client instance with the specified correlation ID
   * 
   * This method allows for easy propagation of correlation IDs across service calls.
   * 
   * @param correlationId Correlation ID to use for requests
   * @returns New client instance with the correlation ID set
   */
  withCorrelationId(correlationId: string): StandardServiceClient<T> {
    // Create a copy of the configuration
    const configCopy = { ...this.config };
    
    // Update headers with correlation ID
    configCopy.headers = {
      ...configCopy.headers,
      'X-Correlation-ID': correlationId
    };
    
    // Create new client with updated configuration
    return new (this.constructor as any)(configCopy);
  }
}

File: common-lib/common_lib/clients/templates/service_client_template.py
"""
Standardized Service Client Template

This module provides a template for creating service clients that follow the
platform's standardized patterns for service communication, error handling,
resilience, and metrics collection.

Usage:
1. Copy this template to your service's clients directory
2. Rename the class to match your service (e.g., MarketDataClient)
3. Implement service-specific methods using the base HTTP methods
4. Create a factory function in your service's client_factory.py

Example:
```python
from common_lib.clients import BaseServiceClient, ClientConfig
from typing import Dict, Any, Optional, Union, List

class ExampleServiceClient(BaseServiceClient):
    \"\"\"Client for interacting with Example Service.\"\"\"
    
    def __init__(self, config: Union[ClientConfig, Dict[str, Any]]):
        \"\"\"Initialize the client.\"\"\"
        super().__init__(config)
    
    async def get_resource(self, resource_id: str) -> Dict[str, Any]:
        \"\"\"Get a resource by ID.\"\"\"
        return await self.get(f"resources/{resource_id}")
```
"""

import logging
from typing import Dict, Any, Optional, Union, List, TypeVar, Generic
from datetime import datetime

from common_lib.clients.base_client import BaseServiceClient, ClientConfig
from common_lib.clients.exceptions import (
    ClientError,
    ClientConnectionError,
    ClientTimeoutError,
    ClientValidationError,
    ClientAuthenticationError
)

# Type variable for response data
T = TypeVar('T')


class StandardServiceClient(BaseServiceClient, Generic[T]):
    """
    Standardized service client template.
    
    This class extends BaseServiceClient with additional standardized methods
    and patterns for service communication.
    
    Features:
    1. Consistent error handling and logging
    2. Type hints for response data
    3. Standardized method signatures
    4. Built-in request/response logging
    5. Correlation ID propagation
    """
    
    def __init__(self, config: Union[ClientConfig, Dict[str, Any]]):
        """
        Initialize the standardized service client.
        
        Args:
            config: Client configuration
        """
        super().__init__(config)
        self.logger = logging.getLogger(f"{__name__}.{self.__class__.__name__}")
    
    async def get_resource(self, resource_id: str) -> Dict[str, Any]:
        """
        Template method for getting a resource by ID.
        
        Args:
            resource_id: Resource identifier
            
        Returns:
            Resource data
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        self.logger.debug(f"Getting resource {resource_id}")
        try:
            return await self.get(f"resources/{resource_id}")
        except Exception as e:
            self.logger.error(f"Failed to get resource {resource_id}: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def create_resource(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Template method for creating a resource.
        
        Args:
            data: Resource data
            
        Returns:
            Created resource data
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        self.logger.debug(f"Creating resource")
        try:
            return await self.post("resources", data=data)
        except Exception as e:
            self.logger.error(f"Failed to create resource: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def update_resource(self, resource_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Template method for updating a resource.
        
        Args:
            resource_id: Resource identifier
            data: Updated resource data
            
        Returns:
            Updated resource data
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        self.logger.debug(f"Updating resource {resource_id}")
        try:
            return await self.put(f"resources/{resource_id}", data=data)
        except Exception as e:
            self.logger.error(f"Failed to update resource {resource_id}: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def delete_resource(self, resource_id: str) -> Dict[str, Any]:
        """
        Template method for deleting a resource.
        
        Args:
            resource_id: Resource identifier
            
        Returns:
            Deletion confirmation
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        self.logger.debug(f"Deleting resource {resource_id}")
        try:
            return await self.delete(f"resources/{resource_id}")
        except Exception as e:
            self.logger.error(f"Failed to delete resource {resource_id}: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def list_resources(
        self, 
        filters: Optional[Dict[str, Any]] = None, 
        page: int = 1, 
        page_size: int = 100
    ) -> Dict[str, Any]:
        """
        Template method for listing resources.
        
        Args:
            filters: Optional filters to apply
            page: Page number
            page_size: Number of items per page
            
        Returns:
            List of resources
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        self.logger.debug(f"Listing resources (page={page}, page_size={page_size})")
        params = {"page": page, "page_size": page_size}
        if filters:
            params.update(filters)
            
        try:
            return await self.get("resources", params=params)
        except Exception as e:
            self.logger.error(f"Failed to list resources: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def execute_operation(
        self, 
        operation: str, 
        resource_id: Optional[str] = None, 
        data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Template method for executing a custom operation.
        
        Args:
            operation: Operation name
            resource_id: Optional resource identifier
            data: Optional operation data
            
        Returns:
            Operation result
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
            ClientValidationError: If the request is invalid
            ClientAuthenticationError: If authentication fails
        """
        endpoint = f"operations/{operation}"
        if resource_id:
            endpoint = f"resources/{resource_id}/{operation}"
            
        self.logger.debug(f"Executing operation {operation}")
        try:
            return await self.post(endpoint, data=data)
        except Exception as e:
            self.logger.error(f"Failed to execute operation {operation}: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def get_health(self) -> Dict[str, Any]:
        """
        Check the health of the service.
        
        Returns:
            Health status
            
        Raises:
            ClientError: If the request fails
            ClientConnectionError: If connection to the service fails
            ClientTimeoutError: If the request times out
        """
        self.logger.debug("Checking service health")
        try:
            return await self.get("health")
        except Exception as e:
            self.logger.error(f"Failed to check service health: {str(e)}")
            # Re-raise the exception to be handled by the caller
            raise
    
    async def with_correlation_id(self, correlation_id: str) -> 'StandardServiceClient[T]':
        """
        Create a new client instance with the specified correlation ID.
        
        This method allows for easy propagation of correlation IDs across service calls.
        
        Args:
            correlation_id: Correlation ID to use for requests
            
        Returns:
            New client instance with the correlation ID set
        """
        # Create a copy of the configuration
        config_dict = self.config.model_dump()
        
        # Update headers with correlation ID
        headers = config_dict.get("default_headers", {})
        headers["X-Correlation-ID"] = correlation_id
        config_dict["default_headers"] = headers
        
        # Create new client with updated configuration
        return self.__class__(config_dict)